numba/cuda/cudadecl.py

```diff
@@ -378,6 +379,39 @@ class Cuda_fp16_cmp(ConcreteTemplate):
     return Cuda_fp16_cmp
 
 
+# Until issue 7863 is resolved, if multiple concrete templates are utilized
+# function resolution will pick the wrong implementation and errors will occur.
+# We avoid this by using ConcreteTempalte for the intrinisics and
+# AbstractTemplate
```

Comment:
Thanks for adding more description. I'm still concerned that it's difficult for the reader to understand the situation because the comment is quite vague and this is a complex issue. In particular my thoughts on the explanation are:

- It's not clear what "multiple concrete templates are utilized" means - I know right now that it's for a single function, but in future I may not remember, and someone else may not know.
- Why does function resolution pick the wrong implementation?
- What errors occur? Failure to compile, incorrect behaviour of compiled code, or just suboptimal performance?
- `ConcreteTemplate` and `intrinsics` are mis-spelled.

I know we're going through a lot of iterations here, but my concern is that this code is going to live for a long time and it should be evident from looking only at the code in `cudadecl.py` what is going on for readers into the future - having to go back and dig through old pull requests and comments is inconvenient enough, but especially so when trying to decipher what is happening and why for such a complex problem that we're dealing with here (especially when it's in general very hard to understand typing and function resolution in Numba at the best of times).

I'd suggest using the following comment to describe the situation - what do you think of it?

```suggestion
# If multiple ConcreteTemplates provide typing for a single function, then
# function resolution will pick the first compatible typing it finds even if it
# involves inserting a cast that would be considered undesirable (in this
# specific case, float16s could be cast to float32s for comparisons).
#
# To work around this, we instead use an AbstractTemplate that implements
# exactly the casting logic that we desire. The AbstractTemplate gets
# considered in preference to ConcreteTemplates during typing.
#
# This is tracked as Issue #7863 (https://github.com/numba/numba/issues/7863) -
# once this is resolved it should be possible to replace this AbstractTemplate
# with a ConcreteTemplate to simplify the logic.
```